File: ai-solutions/04/opus-4.5/python/main.py
from pathlib import Path


def read_input():
    input_path = Path(__file__).parent.parent.parent.parent.parent / "inputs" / "04.txt"
    with open(input_path) as f:
        return [list(line.strip()) for line in f.readlines()]


# 8 directions: N, NE, E, SE, S, SW, W, NW
DIRECTIONS = [(-1, 0), (-1, 1), (0, 1), (1, 1), (1, 0), (1, -1), (0, -1), (-1, -1)]


def count_adjacent_rolls(grid, row, col):
    """Count the number of adjacent paper rolls (@) for a given position."""
    rows, cols = len(grid), len(grid[0])
    count = 0
    for dr, dc in DIRECTIONS:
        nr, nc = row + dr, col + dc
        if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == "@":
            count += 1
    return count


def find_accessible_rolls(grid):
    """Find all rolls that can be accessed (fewer than 4 adjacent rolls)."""
    rows, cols = len(grid), len(grid[0])
    accessible = []
    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == "@" and count_adjacent_rolls(grid, r, c) < 4:
                accessible.append((r, c))
    return accessible


def solve():
    grid = read_input()

    # Part 1: Count initially accessible rolls
    part1 = len(find_accessible_rolls(grid))

    # Part 2: Keep removing accessible rolls until none remain
    working_grid = [row[:] for row in grid]  # Deep copy
    total_removed = 0

    while True:
        accessible = find_accessible_rolls(working_grid)
        if not accessible:
            break

        # Remove all accessible rolls
        for r, c in accessible:
            working_grid[r][c] = "."
        total_removed += len(accessible)

    print(part1)
    print(total_removed)


if __name__ == "__main__":
    solve()

File: ai-solutions/04/opus-4.5/rust/src/main.rs
use std::fs;
use std::path::PathBuf;

// 8 directions: N, NE, E, SE, S, SW, W, NW
const DIRECTIONS: [(i32, i32); 8] = [
    (-1, 0), (-1, 1), (0, 1), (1, 1),
    (1, 0), (1, -1), (0, -1), (-1, -1)
];

fn count_adjacent_rolls(grid: &[Vec<char>], row: usize, col: usize) -> usize {
    let rows = grid.len() as i32;
    let cols = grid[0].len() as i32;
    let mut count = 0;
    
    for (dr, dc) in DIRECTIONS.iter() {
        let nr = row as i32 + dr;
        let nc = col as i32 + dc;
        
        if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
            if grid[nr as usize][nc as usize] == '@' {
                count += 1;
            }
        }
    }
    count
}

fn find_accessible_rolls(grid: &[Vec<char>]) -> Vec<(usize, usize)> {
    let mut accessible = Vec::new();
    
    for (r, row) in grid.iter().enumerate() {
        for (c, &cell) in row.iter().enumerate() {
            if cell == '@' && count_adjacent_rolls(grid, r, c) < 4 {
                accessible.push((r, c));
            }
        }
    }
    accessible
}

fn main() {
    // Build path to input file
    let mut input_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    input_path.push("../../../../inputs/04.txt");
    
    let input = fs::read_to_string(&input_path)
        .expect("Failed to read input file");
    
    // Parse grid
    let grid: Vec<Vec<char>> = input
        .lines()
        .filter(|line| !line.is_empty())
        .map(|line| line.chars().collect())
        .collect();
    
    // Part 1: Count initially accessible rolls
    let part1 = find_accessible_rolls(&grid).len();
    
    // Part 2: Keep removing accessible rolls until none remain
    let mut working_grid = grid.clone();
    let mut total_removed = 0;
    
    loop {
        let accessible = find_accessible_rolls(&working_grid);
        if accessible.is_empty() {
            break;
        }
        
        // Remove all accessible rolls
        for (r, c) in &accessible {
            working_grid[*r][*c] = '.';
        }
        total_removed += accessible.len();
    }
    
    println!("{}", part1);
    println!("{}", total_removed);
}

